### ISSUE TITLE
Incorrect result on function contains adjoint.

### ISSUE THREAD

**Issue by KDr2** [JuliaDiff/ReverseDiff.jl#183]
As mentioned at https://github.com/JuliaDiff/ReverseDiff.jl/pull/180#issuecomment-916868570, when a function contains a call to ``adjoint`, ReverseDiff (with latest dependencies) gives a gradient that differs from other AD packages:

This is an MWE:
```julia
using ForwardDiff, ReverseDiff
using LinearAlgebra: norm

f(x) = sum(x' .* x)
x = rand(5)

test = ForwardDiff.gradient!(DiffResults.GradientResult(x), f, x)
@show DiffResults.gradient(test)

seedx = rand(eltype(x), size(x))
tp = ReverseDiff.GradientTape(f, seedx)
@show ReverseDiff.gradient!(tp, x)
```
Run it on the master branch of ReverseDiff, and ForwardDiff=0.10.18, DiffRules=1.0.2, the two results will be the same.
Run it on the master branch of ReverseDiff, and ForwardDiff=0.10.19, DiffRules=1.3.0, the two results will be different.

The problem was found after DiffRules added these two rules lately:
```julia
@define_diffrule Base.conj(x)                 = :(  1                                  )
@define_diffrule Base.adjoint(x)              = :(  1                                  )
```

https://github.com/JuliaDiff/DiffRules.jl/blob/cbf17ea233d16deb6f0a32bca661c63a79126adc/src/rules.jl#L70

**Comment by devmotion** [comment]
Note that it's not a general problem with `adjoint`, eg it works fine without broadcasting:
```julia
julia> using ForwardDiff, ReverseDiff

julia> f(x) = sum(x' * x);

julia> x = rand(5);

julia> ForwardDiff.gradient(f, x) ≈ 2 * x
true

julia> ReverseDiff.gradient(f, x) ≈ 2 * x
true
```
whereas
```julia
julia> f(x) = sum(x' .* x);

julia> ForwardDiff.gradient(f, x) ≈ fill(2 * sum(x), length(x))
true

julia> ReverseDiff.gradient(f, x) ≈ fill(2 * sum(x), length(x))
false

# but:
julia> ReverseDiff.gradient(f, x) ≈ fill(sum(x), length(x))
true
```

So I think we should not opt out of the `adjoint` rules in DiffRules but use these optimizations and only fix the broken cases with broadcasting.

**Comment by devmotion** [comment]
I guess the first version still works due to this massive specializations for `Adjoint`: https://github.com/JuliaDiff/ReverseDiff.jl/blob/01041c8e8237ed42f6414c6fe0f6e6b12162b6ac/src/derivatives/linalg/arithmetic.jl#L160-L323 :sweat_smile:

I'm sure it's included for a reason but this seems a bit insane and not like a nice solution that generalizes easily to other wrapper types.

**Comment by devmotion** [comment]
Another observation:

The same problem exists for `transpose`, and it existed already for DiffRules 1.0.2 (which defined the same rule for `transpose`):
```julia
pkg> add DiffRules@1.0.2, ReverseDiff

julia> using ReverseDiff

julia> x = rand(5);

julia> f(x) = sum(x' .* x);

julia> ReverseDiff.gradient(f, x) ≈ fill(2 * sum(x), length(x))
true

julia> ReverseDiff.ForwardDiff.gradient(f, x) ≈ fill(2 * sum(x), length(x))
true

julia> f(x) = sum(transpose(x) .* x);

julia> ReverseDiff.gradient(f, x) ≈ fill(2 * sum(x), length(x))
false

julia> ReverseDiff.ForwardDiff.gradient(f, x) ≈ fill(2 * sum(x), length(x))
true
```

**Comment by devmotion** [comment]
I think I figured out the main underlying problem here. And a simple fix seems to be:
```julia
julia> using ReverseDiff, LinearAlgebra

julia> Base.@propagate_inbounds Base.getindex(v::LinearAlgebra.AdjOrTransAbsVec{<:ReverseDiff.TrackedReal}, i::Int) = v.parent[i-1+first(axes(v.parent)[1])]

julia> f(x) = sum(x' .* x);

julia> x = rand(5);

julia> ReverseDiff.gradient(f, x) ≈ fill(2 * sum(x), length(x))
true

julia> f(x) = sum(transpose(x) .* x);

julia> x = rand(5);

julia> ReverseDiff.gradient(f, x) ≈ fill(2 * sum(x), length(x))
true
```

**Comment by devmotion** [comment]
A more detailed explanation:
- The `Adjoint{<:TrackedReal}` and `Transpose{<:TrackedReal}` correctly hit https://github.com/JuliaDiff/ReverseDiff.jl/blob/01041c8e8237ed42f6414c6fe0f6e6b12162b6ac/src/derivatives/propagation.jl#L35 (and similarly for `decrement_deriv!`)
- Indexing into such an array calls https://github.com/JuliaLang/julia/blob/c5f348726cebbe55e169d4d62225c2b1e587f497/stdlib/LinearAlgebra/src/adjtrans.jl#L178 and lines below
- **Issue**: The definition in LinearAlgebra calls `adjoint`/`transpose`  (`wrapperop(v)`) on the retrieved/inserted values but the new rule for `adjoint` and the old one for `transpose` lead to a `TrackedReal` with a derivative of zero!

So it seems to be sufficient to implement https://github.com/JuliaLang/julia/blob/c5f348726cebbe55e169d4d62225c2b1e587f497/stdlib/LinearAlgebra/src/adjtrans.jl#L178-L181 (for completeness probably all `getindex`/`setindex!` methods should be implemented) for `T<:TrackedReal` without `wrapperop(v)`, i.e.
```julia
Base.@propagate_inbounds Base.getindex(v::LinearAlgebra.AdjOrTransAbsVec{<:TrackedReal}, i::Int) = v.parent[i-1+first(axes(v.parent)[1])]
Base.@propagate_inbounds Base.getindex(A::LinearAlgebra.AdjOrTransAbsMat{<:TrackedReal}, i::Int, j::Int) = A.parent[j, i]
Base.@propagate_inbounds Base.setindex!(v::LinearAlgebra.AdjOrTransAbsVec{<:TrackedReal}, x, i::Int) = setindex!(v.parent, x, i-1+first(axes(v.parent)[1])); v)
Base.@propagate_inbounds Base.setindex!(A::LinearAlgebra.AdjOrTransAbsMat{<:TrackedReal}, x, i::Int, j::Int) = (setindex!(A.parent, x, j, i); A)
```

**Comment by devmotion** [comment]
On a second thought, probably it is easier and leads to fewer problems (no inconsistencies, does not become outdated) if we just rely on the defaults in https://github.com/JuliaLang/julia/blob/c5f348726cebbe55e169d4d62225c2b1e587f497/base/number.jl#L211-L213 and the rules for `transpose`, `adjoint` and `conj` are removed from DiffRules. This was already done for `identity` since it caused problems: https://github.com/JuliaDiff/DiffRules.jl/pull/64

**Comment by devmotion** [comment]
Fixed by DiffRules 1.3.1.
